spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/BufferedIterator2.scala

```diff
@@ -0,0 +1,56 @@
+package com.datastax.spark.connector.util
+
+import scala.collection.mutable.ArrayBuffer
+
+/** Serves the same purpose as `BufferedIterator` in Scala, but its `takeWhile` method 
+  * properly doesn't consume the next element. */
+class BufferedIterator2[T](iterator: Iterator[T]) extends Iterator[T] {
+
+  // Instead of a pair of T and Boolean we could use Option, but
+  // this would allocate a new Option object per each item, which might be
+  // too big overhead in tight loops using this iterator.
+  private[this] var headDefined: Boolean = false
```

Comment:
Personally I would not use this and would prefer to see

    if (iterator.hasNext) advance

It is clearer and removes the need for a var